### ISSUE TITLE
Plugin API

### ISSUE THREAD

**Issue by Artiom-M** [Ribtoks/xpiks#182]
I have some side tools for picture analysis and it would be good to integrate them to the application. However these tools are quite complex and CPU consuming, so integrate them as is looks like not a good idea. How is complicated to create a plugin API for external integration?
For example I have two useful tools:
1. blur detection - it can give you a hint similar to insufficient resolution or missing keywords.
2. Inventory tool - this one is mostly a database, keeping sync between local photo library and remote IDs (accepted photos). I would like to extend it with "uploaded" status when xpiks successfully uploaded a photo and in xpiks show some information from this database.

**Comment by Ribtoks** [comment]
It's doable. For sure. But I have another idea.
Plugin development would be a lot more painful then, say, service-like architecture for desktop. Pros of services are that if they will crash, Xpiks will still be working - just without them. If plugins (dlls) will crash, Xpiks will crash as well, and I'll have to have a lot of try-catch items.

For example, Xpiks can have a protocol for Warnings Checking: (registering a Checker, submitting item for checking, receiving a result). Your service will register itself in Xpiks as a protocol implementation and Xpiks can submit items to Warnings Checker when checking for warnings through the protocol to your service as well.

**Comment by Ribtoks** [comment]
Writing a Plugin API will require more time. Not the API itself (it could be done in several days), but it's integration with everything in Xpiks - it may take a lot of time.

**Comment by Artiom-M** [comment]
I prefer service approach. I choose "Plugin API" term as most understandable for me.
So Warnings Checker is a good start.

**Comment by Ribtoks** [comment]
Yes, I prefer it as well. In future it will enable to run your magical service on something like Amazon Lambda.

**Comment by Ribtoks** [comment]
Also before developing a plugin system, it's worth adding more tests to Xpiks not to f&ck up everything. See #75

**Comment by Ribtoks** [comment]
After I spent week implementing client-server, I found articles about how Qt handles it (via QPluginLoader) and I reimplemented it to be native to Qt. The API itself of course will change, adapt and improve in the following months, while I'll be implementing some other plugins as well.

Comprehensive example of how to use plugin API basics are in the [Hello World Plugin](https://github.com/ribtoks/xpiks-helloworld-plugin) repository in my account. Examples include my recommendation on implementing Worker-like types of plugins, plugins with UI and plugins with registered menu items in Xpiks.

Don't hesitate to contact me in case you will have any questions of how best to use the API itself. Of course, I will amend it with the abilities you will need for you plugins. Also there would be an article in the Xpiks Wiki with instructions on plugin development.
